jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/DemandingFlusher.java

```diff
@@ -0,0 +1,181 @@
+     */
+    public DemandingFlusher(BiConsumer<Frame, Callback> emitFrame)
+    {
+        _emitFrame = emitFrame;
+    }
+
+    /**
+     * <p>Called when there is demand for a single frame to be produced. During this method a single call can be made
+     * to {@link #emitFrame(Frame, Callback)} which will forward this frame towards the application. Returning true
+     * from this method signals that you are done processing the current Frame, and the next invocation of this
+     * method will have the next frame.</p>
+     *
+     * <p>Note that the callback supplied here is specially wrapped so that you can call
+     * it multiple times and it will not be completed more than once. This simplifies the
+     * handling of failure cases.</p>
+     * @param frame the original frame.
+     * @param callback to succeed to release the frame payload.
+     * @param first if this is the first time this method has been called for this frame.
+     * @return false to continue processing this frame, true to complete processing and get a new frame.
+     */
+    protected abstract boolean handle(Frame frame, Callback callback, boolean first);
+
+    @Override
+    public void demand(long n)
+    {
+        _demand.getAndUpdate(d -> Math.addExact(d, n));
+        iterate();
+    }
+
+    @Override
+    public void setNextDemand(LongConsumer nextDemand)
+    {
+        _nextDemand = nextDemand;
+    }
+
+    /**
+     * Used to supply the flusher with a new frame. This frame should only arrive if demanded
+     * through the {@link LongConsumer} provided by {@link #setNextDemand(LongConsumer)}.
+     * @param frame the WebSocket frame.
+     * @param callback to release frame payload.
+     */
+    public void onFrame(Frame frame, Callback callback)
+    {
+        if (_frame != null || _callback != null)
+            throw new IllegalStateException("Not expecting onFrame");
+
+        _frame = frame;
+        _callback = new CountingCallback(callback, 1);
+        succeeded();
+    }
+
+    /**
+     * Used to fail this flusher possibly from an external event such as a callback.
+     * @param t the failure.
+     */
+    public void failFlusher(Throwable t)
+    {
+        if (_failure.compareAndSet(null, t))
+        {
+            failed(t);
+            iterate();
+        }
+    }
+
+    /**
+     * <p>This is used within an implementation of {@link #handle(Frame, Callback, boolean)}
+     * to forward a frame onto the next layer of processing.</p>
+     *
+     * <p>This method should only be called ONCE within each invocation of {@link #handle(Frame, Callback, boolean)}
+     * otherwise</p>
+     * @param frame the WebSocket frame.
+     * @param callback to release frame payload.
+     */
+    public void emitFrame(Frame frame, Callback callback)
+    {
+        if (_demand.decrementAndGet() < 0)
+            throw new IllegalStateException("Negative Demand");
```

Comment:
Do we need to increment demand again?

Reply:
The increment is done inside the call to demand that we intercept.

Reply:
Hmm I get what you mean, the application has already demanded, but we have thrown ISE but don't have demand for the close Frame.

But I don't think this is a big issue because we don't wait for close response if we send an error close code.

And this ISE can only be trigged by a bad implementation of the `DemandingFlusher`. In the javadoc it said to only call `emitFrame` once per call to `handle()`.

Reply:
yeah but no!   Ideally an implementation would receive an ISE in response to a bad call, that would not actually change the state of the flusher.     But kind of difficult to achieve atomically without a CAS loop:
```suggestion
        while(true)
        {
            int demand = _demand.get();
            if (demand <= 0)
                throw new IllegalStateException("Negative Demand");
           if (_demand.compareAndSet(demand, demand - 1)
               break;
       }
```